File: src/main/java/leetcode/lists/hot200/LeetCode285.java
package leetcode.lists.hot200;

import common.*;

import java.util.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/3/12 16:13
 */
public class LeetCode285 {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if (root == null) {
            return null;
        }
        Stack<TreeNode> stack = new Stack<>();
        TreeNode prev = null;
        while (!stack.isEmpty() || root != null) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            if (prev == p) {
                return root;
            }
            prev = root;
            root = root.right;
        }
        return null;
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode285.go
package topinterview

func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {

	if root == nil {
		return nil
	}

	var dfs func(root *TreeNode) *TreeNode
	var flag bool
	var res *TreeNode
	dfs = func(root *TreeNode) *TreeNode {
		if root == nil {
			return nil
		}
		dfs(root.Left)
		if flag {
			res = root
			flag = !flag
		}
		if root == p {
			flag = true
		}
		dfs(root.Right)
		return res
	}
	return dfs(root)
}
